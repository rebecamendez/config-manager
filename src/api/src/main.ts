import { NestFactory } from '@nestjs/core';
import { AppModule } from './app.module';
import { Logger } from '@nestjs/common';

async function bootstrap(): Promise<void> {
  const port = 3000;
  const app = await NestFactory.create(AppModule);
  await app.listen(port);
  Logger.log(`ðŸŽ§ ConfigManager API start listening on the port: ${port}`);
}

void bootstrap();
