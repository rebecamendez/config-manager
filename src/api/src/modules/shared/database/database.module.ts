import { Module } from '@nestjs/common';
import { databaseProviders, DATA_SOURCE } from './database.providers';

@Module({
  providers: [...databaseProviders],
  exports: [DATA_SOURCE]
})
export class DatabaseModule {}
