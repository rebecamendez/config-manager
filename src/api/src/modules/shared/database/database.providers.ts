import { DataSource } from 'typeorm';
import { typeormDataSource } from './database.typeorm.datasource';

export const DATA_SOURCE = 'DATA_SOURCE';

export const databaseProviders = [
  {
    provide: DATA_SOURCE,
    useFactory: async (): Promise<DataSource> => {
      return typeormDataSource.initialize();
    }
  }
];
