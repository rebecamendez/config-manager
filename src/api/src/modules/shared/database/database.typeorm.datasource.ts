import { ConfigService } from '@nestjs/config';
import { DataSource } from 'typeorm';
import { config } from 'dotenv';

config();

const configService = new ConfigService();

// https://stackoverflow.com/questions/61259812/cannot-use-import-statement-outside-a-module-in-typeorm-migration-when-run-nes
export const typeormDataSource = new DataSource({
  type: 'postgres',
  host: configService.getOrThrow<string>('DB_HOST'),
  port: 5432,
  username: configService.getOrThrow<string>('DB_USER'),
  password: configService.getOrThrow<string>('DB_PASSWORD'),
  database: configService.getOrThrow<string>('DB_DATABASE'),
  entities: [__dirname + '/entities/**/*.entity{.ts,.js}'],
  migrations: ['./dist/src/migrations/**/*.js'],
  migrationsRun: false,
  synchronize: false,
  logging: configService.getOrThrow<boolean>('DB_ENABLE_LOGGER')
});
