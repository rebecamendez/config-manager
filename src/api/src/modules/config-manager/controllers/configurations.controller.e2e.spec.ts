/* eslint-disable no-process-env */
import { INestApplication } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import * as request from 'supertest';
import { AppModule } from '../../../app.module';
import { setupNestApp } from '../../../main.setup';
import { GenericContainer, StartedTestContainer } from 'testcontainers';
import { DATA_SOURCE } from 'modules/shared/database/database.providers';
import { DataSource } from 'typeorm';
import { createMockedConfiguration } from '../tests/mocks/configurations.mocks';

describe('A Configurations controller (e2e)', () => {
  let app: INestApplication;
  let container: StartedTestContainer;

  beforeAll(async () => {
    const databaseName = 'configurations-controller-e2e-test';
    const entities = [__dirname + './../../shared/database/entities/**/*.entity.ts'];

    container = await new GenericContainer('postgres')
      .withEnvironment({ POSTGRES_USER: 'the-user', POSTGRES_PASSWORD: 'the-password', POSTGRES_DB: databaseName })
      .withExposedPorts(5432)
      .withCopyFilesToContainer([{ source: 'init.sql', target: '/docker-entrypoint-initdb.d/init.sql' }])
      .withName(databaseName)
      .start();

    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule]
    })
      .overrideProvider(DATA_SOURCE)
      .useFactory({
        factory: async () => {
          const datasourceConfig = new DataSource({
            type: 'postgres',
            host: container.getHost(),
            port: container.getMappedPort(5432),
            username: 'the-user',
            password: 'the-password',
            database: databaseName,
            entities: entities,
            migrations: ['./dist/src/migrations/**/*.js'],
            migrationsRun: false,
            synchronize: false,
            logging: false
          });
          return datasourceConfig.initialize();
        }
      })
      .compile();

    app = moduleFixture.createNestApplication();
    setupNestApp(app);
    await app.init();
  });

  afterAll(async () => {
    if (app) await app.close();
    if (container) await container.stop();
  });

  it('should return configurations', async () => {
    const response = await request(app.getHttpServer()).get('/configurations').send();
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });

  it('should return a configuration', async () => {
    const response = await request(app.getHttpServer()).get('/configurations/my-config').send();
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });

  it('should throw an 404 error trying to retrieve a non existent config', async () => {
    const response = await request(app.getHttpServer()).get('/configurations/my-configNotFound').send();
    expect(response.status).toBe(404);
  });

  it('should create a configuration', async () => {
    const configuration = createMockedConfiguration();
    const response = await request(app.getHttpServer()).post('/configurations').send(configuration);
    expect(response.status).toBe(201);
    expect(response.body).toMatchSnapshot();
  });

  it('should throw an 409 error trying to create an existant key config', async () => {
    const configuration = createMockedConfiguration({ key: 'my-config' });
    const response = await request(app.getHttpServer()).post('/configurations').send(configuration);
    expect(response.status).toBe(409);
  });
});
