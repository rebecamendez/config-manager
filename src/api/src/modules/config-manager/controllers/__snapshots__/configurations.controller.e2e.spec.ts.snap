// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Configurations controller (e2e) GET: should return a configuration 1`] = `
{
  "key": "my-config",
  "value": "myValue",
}
`;

exports[`A Configurations controller (e2e) GET: should return configurations 1`] = `
[
  {
    "key": "my-config",
    "value": "myValue",
  },
]
`;

exports[`A Configurations controller (e2e) POST: should create a configuration 1`] = `
{
  "key": "foo",
  "value": "bar",
}
`;

exports[`A Configurations controller (e2e) PUT: should upate a configuration 1`] = `
{
  "key": "my-config",
  "value": "updated value",
}
`;
